scipy/spatial/_hausdorff.pyx

```diff
@@ -0,0 +1,63 @@
+"""
+Directed Hausdorff Code
+
+.. versionadded:: 0.19.0
+
+"""
+#
+# Copyright (C)  Tyler Reddy and Richard Gowers, 2016
+#
+# Distributed under the same BSD license as Scipy.
+#
+
+import numpy as np
+cimport numpy as np
+cimport cython
+from libc.math cimport sqrt
+
+__all__ = ['directed_hausdorff']
+
+@cython.boundscheck(False)
+def directed_hausdorff(double[:,::1] ar1, double[:,::1] ar2):
+
+    cdef double cmax, cmin, d
+    cdef int break_occurred
+    cdef int N1 = ar1.shape[0]
+    cdef int N2 = ar2.shape[0]
+    cdef int data_dims = ar1.shape[1]
+    cdef unsigned int i, j, k, i_store, j_store, i_ret, j_ret
+    cdef long[:] resort1, resort2
+
+    # shuffling the points in each array generally increases the likelihood of
+    # an advantageous break in the inner search loop and never decreases the
+    # performance of the algorithm
+    np.random.seed(123)
+    np.random.shuffle(np.asarray(ar1))
+    np.random.shuffle(np.asarray(ar2))
+    resort1 = np.arange(N1)
+    resort2 = np.arange(N2)
+    np.random.shuffle(np.asarray(resort1))
+    np.random.shuffle(np.asarray(resort2))
```

Comment:
I considered this as well. I will look into the option of avoiding random seed fixing to avoid my (and your) reservations.

I'm not sure I understand what you're saying about the code proper though. Below you argue that my function is effectively altering the global random state in a way that the end user may not be aware of, but here you're concerned that the random state may change within the function itself? If you fix the random seed, it will ALWAYS shuffle an array of the same size in the same way, so you can restore the original order of elements by shuffling a numpy array of indices and then looking up that array at the end.